### PR TITLE
Implemented Disjoint Set (Union-Find) Data Structure

### PR DESCRIPTION
**Pull request by Ramy-Badr-Ahmed**
**Contents:**

- `DisjointSetNode` Class: Represents individual nodes with `data`, `rank`, and `parent` attributes.

- DisjointSet Class: Implements core methods:

    `findSet`: Utilises path compression for representative lookup.
    `unionSet`: Applies union by rank to manage the merging of sets.

- `Unit Tests`: Tests using PHPUnit to validate the correct behavior of the Disjoint Set structure. The tests cover different numbers of nodes and sets.

**Compatibility**: 
The implementation supports PHP 7.4 and above by omitting the `mixed` type hint, otherwise replace with `mixed` type hint as of `PHP 8.0^`.

### CODE REVIEW

**Comment by Ramy-Badr-Ahmed** [comment]
@darwinz 
Looking forward for your review :-)

**Comment by Ramy-Badr-Ahmed** [comment]
@darwinz 
Looking forward for your review :-)

**Comment by Md-Kasali** [comment]
Move your DisjointSetTest.php file under tests/DataStructures/ directory

**Comment by Md-Kasali** [comment]
Move your DisjointSetTest.php file under the DataStructures/ directory into the tests/DataStructures directory. The test files are organized under the tests/ directory.

**Review by Ramy-Badr-Ahmed — COMMENTED** [review]
Moved `DisjointSetTest.php` to  `tests/DataStructures`

File: DataStructures/DisjointSets/DisjointSetNode.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace DataStructures\DisjointSets;
+
+class DisjointSetNode
+{
+    public mixed $data;     # PHP 8.0^ . Otherwise, define with annotations:  @var int|string|float|null
+    public int $rank;
+    public DisjointSetNode $parent;
+
+    public function __construct(mixed $data = NULL) {
```
Comment (darwinz) [inline DataStructures/DisjointSets/DisjointSetNode.php:11]:
```suggestion
    public function __construct(mixed $data = NULL) 
    {
```

File: DataStructures/DisjointSets/DisjointSetNode.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        $this->data = $data;
+        $this->rank = 0;
+        $this->parent = $this;  // Initialize parent to itself
+    }
+}
```
Comment (darwinz) [inline DataStructures/DisjointSets/DisjointSetNode.php:16]:
```suggestion
}

```

File: DataStructures/DisjointSets/DisjointSet.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        if ($rootX->rank > $rootY->rank) {
+            $rootY->parent = $rootX;
+        } else {
+            $rootX->parent = $rootY;
+            if ($rootX->rank === $rootY->rank) {
+                $rootY->rank += 1;
+            }
+        }
+    }
+}
```
Comment (darwinz) [inline DataStructures/DisjointSets/DisjointSet.php:40]:
```suggestion
}

```

File: tests/DataStructures/DisjointSetTest.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        $this->ds->unionSet($this->nodes[12], $this->nodes[13]);
+        $this->ds->unionSet($this->nodes[13], $this->nodes[14]);
+
+        $this->ds->unionSet($this->nodes[15], $this->nodes[16]);
+        $this->ds->unionSet($this->nodes[16], $this->nodes[17]);
+
+        $this->ds->unionSet($this->nodes[18], $this->nodes[19]);
+
+    }
```
Comment (darwinz) [inline tests/DataStructures/DisjointSetTest.php:45]:
```suggestion
    }
```

File: tests/DataStructures/DisjointSetTest.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            [6, 7, 8],      // Set C
+            [9, 10, 11],    // Set D
+            [12, 13, 14],   // Set E
+            [15, 16, 17],   // Set F
+            [18, 19]        // Set G
+        ];
+
+        foreach ($sets as $set) {
+            if (in_array($i, $set) && in_array($j, $set))
+                return true;
```
Comment (darwinz) [inline tests/DataStructures/DisjointSetTest.php:79]:
```suggestion
            if (in_array($i, $set) && in_array($j, $set)) {
                return true;
            }    
```

File: tests/DataStructures/DisjointSetTest.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        return false;
+    }
+
+}
```
Comment (darwinz) [inline tests/DataStructures/DisjointSetTest.php:85]:
```suggestion
}

```

**Review by darwinz — CHANGES_REQUESTED** [review]
This is great, thanks for the contribution!

There are a few small linting issues, found by the GitHub Actions linter we use in this repo.  You can also run PHPCBF to automatically fix these types of issues in the future.

**Comment by darwinz** [comment]
PHPUnit is also failing, so you'll want to fix that error as well.

```
1) DataStructures\DisjointSetTest::testFindSet
Error: Class 'DisjointSet' not found

/home/runner/work/PHP/PHP/tests/DataStructures/DisjointSetTest.php:16
phpvfscomposer:///home/runner/work/PHP/PHP/vendor/phpunit/phpunit/phpunit:106
```

**Comment by Ramy-Badr-Ahmed** [comment]
> This is great, thanks for the contribution!
> 
> There are a few small linting issues, found by the GitHub Actions linter we use in this repo. You can also run PHPCBF to automatically fix these types of issues in the future.

Thanks!
I have committed your changes and considered the rest of the PHPCP remarks.

**Comment by Ramy-Badr-Ahmed** [comment]
> PHPUnit is also failing, so you'll want to fix that error as well.
> 
> ```
> 1) DataStructures\DisjointSetTest::testFindSet
> Error: Class 'DisjointSet' not found
> 
> /home/runner/work/PHP/PHP/tests/DataStructures/DisjointSetTest.php:16
> phpvfscomposer:///home/runner/work/PHP/PHP/vendor/phpunit/phpunit/phpunit:106
> ```

It works now with a corrected namespace and manual `requires` following the exisitng pattern 🙂

```php
Testing started at 13:48 ...
PHPUnit 9.6.20 by Sebastian Bergmann and contributors.

Time: 00:00.004, Memory: 4.00 MB

OK (1 test, 36 assertions)

Process finished with exit code 0
```

**Comment by darwinz** [comment]
The test is still failing.  Please run your test locally and make sure it's passing.

**Comment by Ramy-Badr-Ahmed** [comment]
Hi Brandon,

Thanks for the info. I have resolved the issues and Github actions have passed in my forked version. Can you please check it out again?

Thanks a lot for your time

Kind regards,
Ramy

**Review by darwinz — APPROVED** [review]

**Comment by darwinz** [comment]
Thank you!

**Pull request merged**